elyra/pipeline/component_catalog.py

```diff
@@ -25,25 +29,129 @@
+from traitlets.config import SingletonConfigurable
 
 from elyra.metadata.manager import MetadataManager
-from elyra.metadata.metadata import Metadata
 from elyra.metadata.schemaspaces import ComponentCatalogs
+from elyra.pipeline.catalog_connector import ComponentCatalogConnector
 from elyra.pipeline.component import Component
 from elyra.pipeline.component import ComponentParser
+from elyra.pipeline.component_metadata import ComponentCatalogMetadata
+from elyra.pipeline.runtime_type import RuntimeProcessorType
 
+BLOCKING_TIMEOUT = 5
+NONBLOCKING_TIMEOUT = 0.10
+CATALOG_UPDATE_TIMEOUT = 15
 
-class ComponentCatalog(LoggingConfigurable):
-    """
-    Component Catalog, responsible to provide a list of available components
-    for each runtime. The catalog uses component parser to read and parse each
-    component entry from the catalog and transform them into a component value object.
-    """
-    _cached_components: Dict[str, Component] = {}
-    _cache_last_updated = None
+
+class CacheUpdateManagerThread(Thread):
+    def __init__(self, log: Logger, component_cache: Dict[str, Dict], cache_task_queue: Queue):
+        super().__init__()
+
+        self.setDaemon(True)
+
+        self.log = log
+        self._component_cache = component_cache
+        self._cache_task_queue = cache_task_queue
+
+        self._threads: List[CacheUpdateThread] = []
+
+        self.stop_event = Event()
+
+    def run(self):
+        outstanding_threads = False
+
+        while not self.stop_event.is_set():
+            timeout = NONBLOCKING_TIMEOUT if outstanding_threads else BLOCKING_TIMEOUT
+            try:
+                # Get a task from the queue
+                catalog, operation = self._cache_task_queue.get(timeout=timeout)
+
+            except Empty:
+                # No task exists in the queue, proceed to check for thread execution
+                pass
+
+            else:
+                # Create and start a thread for the task
+                updater_thread = CacheUpdateThread(self._component_cache, self._cache_task_queue, catalog, operation)
+                updater_thread.start()
+
+                self._threads.append(updater_thread)
+
+            outstanding_threads = self.check_outstanding_threads()
+
+    def check_outstanding_threads(self) -> bool:
+        """
+        Join finished threads and report on long-running threads as needed.
+        """
+        outstanding_threads = False
+        for thread in self._threads:
+            try:
+                # Attempt to join thread within the given amount of time
+                thread.join(timeout=NONBLOCKING_TIMEOUT)
+
+            except Exception:
+                # Thread is still running (thread join timed out)
+                outstanding_threads = True
+                if time.time() - thread.task_start_time > CATALOG_UPDATE_TIMEOUT:
+                    self.log.warning(f"Thread {thread.name} is taking too long...")  # TODO Fix this message
```

Comment:
Fix this TODO (suggestions welcome here)

Reply:
I would suggest setting the thread's name to the catalog associated with the update by adding `name=catalog.name` when initializing the superclass, then use a message something like...
```python
self.log.warning(f"Cache update for catalog '{thread.name}' is still processing after {duration} seconds ...")
```
where `durantion = int(time.time() - thread.task_start_time)` converted to seconds.  

More importantly though, once the `CATALOG_UPDATE_TIMEOUT` has been initially exceeded, this warning will spam the log at sub-seconds intervals, so this should be prevented.  Ideally, it would be best to warn only at `CATALOG_UPDATE_TIMEOUT` intervals, so one approach would be to add a `last_warn_time` field that is initialized to `task_start_time` when the thread is created, then use that field to determine the warning and reset that field on each logged warning.  

Since catalog readers tend to just execute on a finite set of items, I'm not sure we need to plumb a "give up" mechanism - or can probably hold off on that.  If we did need that, we'd probably also need a counter to track the times we log a warning - which is roughly every `CATALOG_UPDATE_TIMEOUT` seconds.  We'd then pick a value at which _that_ multiple of `CATALOG_UPDATE_TIMEOUT` is the "give up" point.  For example, since `CATALOG_UPDATE_TIMEOUT` is 15 seconds, a count of 20 would be roughly 5 minutes of waiting for that thread to complete - but I don't think we need this right now.

Reply:
Great point and idea! As I was going through to test these updates after making them, I realized that `thread.join` doesn't actually throw an exception, so that has also been fixed in the latest commit and is now working